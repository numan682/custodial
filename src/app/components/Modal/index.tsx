import { createContext, useCallback, useContext, useState } from 'react'
import { Modal, ModalContainer } from './ModalContainer'

interface ModalContextProps {
  launchModal: (modal: Modal) => void
  closeModal: () => void
}

interface ModalProviderProps {
  children: React.ReactNode
}

const ModalContext = createContext<ModalContextProps>({} as ModalContextProps)

export const ModalProvider = (props: ModalProviderProps) => {
  const [modal, setModal] = useState<Modal | null>(null)
  const closeModal = useCallback(() => {
    setModal(null)
  }, [])

  return (
    <ModalContext.Provider value={{ closeModal, launchModal: setModal }}>
      {props.children}
      {modal && <ModalContainer modal={modal} closeModal={closeModal} />}
    </ModalContext.Provider>
  )
}

// eslint-disable-next-line react-refresh/only-export-components
export const useModal = () => {
  const context = useContext(ModalContext)
  if (context === undefined) {
    throw new Error('useModal must be used within a ModalProvider')
  }

  return context
}
