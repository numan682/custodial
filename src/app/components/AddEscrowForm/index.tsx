/**
 *
 * AddEscrowForm
 *
 */
import { useModal } from 'app/components/Modal'
import { parseRoseStringToBaseUnitString } from 'app/lib/helpers'
import { selectMinStaking } from 'app/state/network/selectors'
import { Validator } from 'app/state/staking/types'
import { transactionActions } from 'app/state/transaction'
import { selectTransaction } from 'app/state/transaction/selectors'
import { Box } from 'grommet/es6/components/Box'
import { Button } from 'grommet/es6/components/Button'
import { CheckBox } from 'grommet/es6/components/CheckBox'
import { Form } from 'grommet/es6/components/Form'
import { TextInput } from 'grommet/es6/components/TextInput'
import React, { memo, useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { useDispatch, useSelector } from 'react-redux'

import { TransactionStatus } from '../TransactionStatus'

interface Props {
  validatorAddress: string
  validatorStatus: Validator['status']
  validatorRank: number
}

export const AddEscrowForm = memo((props: Props) => {
  const { t } = useTranslation()
  const { launchModal } = useModal()
  const { error, success } = useSelector(selectTransaction)
  const isTop20 = props.validatorRank <= 20
  const [showNotice, setShowNotice] = useState(isTop20)
  const [amount, setAmount] = useState('')
  const dispatch = useDispatch()
  const minStaking = useSelector(selectMinStaking)

  const delegate = () => {
    dispatch(
      transactionActions.addEscrow({
        type: 'addEscrow',
        amount: parseRoseStringToBaseUnitString(amount),
        validator: props.validatorAddress,
      }),
    )
  }
  const submit = () => {
    if (props.validatorStatus !== 'active') {
      launchModal({
        title: t('account.addEscrow.confirmDelegatingToInactive.title', 'Are you sure you want to continue?'),
        description: t(
          'account.addEscrow.confirmDelegatingToInactive.description',
          'Status of this validator is {{validatorStatus}}. Your delegation might not generate any rewards.',
          { validatorStatus: props.validatorStatus },
        ),
        handleConfirm: delegate,
        isDangerous: true,
      })
    } else {
      delegate()
    }
  }

  useEffect(() => {
    return () => {
      dispatch(transactionActions.clearTransaction())
    }
  }, [dispatch])

  return (
    <>
      {showNotice && (
        <Box pad={{ vertical: 'medium' }} gap="xsmall">
          {t(
            'account.addEscrow.confirmDelegatingToTop.description',
            'This validator is ranked in the top 20 by stake. Please consider delegating to a smaller validator to increase network security and decentralization.',
          )}
          <CheckBox
            label={t('account.addEscrow.confirmDelegatingToTop.acknowledge', 'Delegate anyway')}
            checked={!showNotice}
            onChange={event => setShowNotice(!event.target.checked)}
          />
        </Box>
      )}
      {!showNotice && (
        <Form onSubmit={submit}>
          <Box direction="row" gap="small" pad={{ top: 'small' }}>
            <Box background="background-front">
              <TextInput
                data-testid="amount"
                id="amount-id"
                name="amount"
                placeholder={t('common.amount')}
                type="number"
                step="any"
                min={minStaking}
                value={amount}
                onChange={event => setAmount(event.target.value)}
                required
              />
            </Box>
            <Button label={t('account.addEscrow.delegate', 'Delegate')} type="submit" primary />
          </Box>
          <TransactionStatus error={error} success={success} />
        </Form>
      )}
    </>
  )
})
