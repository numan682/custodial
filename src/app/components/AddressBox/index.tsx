/**
 *
 * AddressBox
 *
 */
import copy from 'copy-to-clipboard'
import { Box } from 'grommet/es6/components/Box'
import { Button } from 'grommet/es6/components/Button'
import { Text } from 'grommet/es6/components/Text'
import { Notification } from 'grommet/es6/components/Notification'
import { Copy } from 'grommet-icons/es6/icons/Copy'
import React, { memo, useState } from 'react'
import { useTranslation } from 'react-i18next'

import { PrettyAddress } from '../PrettyAddress'

interface Props {
  address: string
  border?: boolean
}

export const AddressBox = memo((props: Props) => {
  const { t } = useTranslation()
  const [notificationVisible, setNotificationVisible] = useState(false)

  const hideNotification = () => setNotificationVisible(false)

  const address = props.address

  const copyAddress = () => {
    const wasCopied = copy(address)
    if (wasCopied) {
      setNotificationVisible(true)
    }
  }

  return (
    <Box
      direction="row"
      align="center"
      round="5px"
      pad={{ right: 'small' }}
      width="fit-content"
      border={props.border && { color: 'brand' }}
    >
      <Button onClick={() => copyAddress()} icon={<Copy size="18px" />} data-testid="copy-address" />
      <Text weight="bold" size="medium" wordBreak="break-word">
        <PrettyAddress address={address} />
      </Text>
      {notificationVisible && (
        <Notification
          toast
          status={'normal'}
          title={t('account.addressCopied', 'Address copied.')}
          onClose={hideNotification}
        />
      )}
    </Box>
  )
})
