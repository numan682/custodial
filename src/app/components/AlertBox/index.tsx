import { Box } from 'grommet/es6/components/Box'
import { Text } from 'grommet/es6/components/Text'
import { normalizeColor } from 'grommet/es6/utils'
import * as React from 'react'
import { useContext } from 'react'
import { ThemeContext } from 'styled-components'

export type AlertBoxStatus = 'error' | 'warning' | 'ok' | 'ok-weak' | 'info'

interface Props {
  status: AlertBoxStatus
  center?: boolean
  /** Example `icon={<Info size="20px" color="currentColor" />}` */
  icon?: React.ReactNode
  children: React.ReactNode
}

const mapStatus = {
  error: {
    color: 'alert-box-error',
    background: 'alert-box-error-background',
  },
  warning: {
    color: 'alert-box-warning',
    background: 'alert-box-warning-background',
  },
  ok: {
    color: 'alert-box-ok',
    background: 'alert-box-ok-background',
  },
  'ok-weak': {
    color: 'alert-box-ok-weak',
    background: 'alert-box-ok-weak-background',
  },
  info: {
    color: 'alert-box-info',
    background: 'alert-box-info-background',
  },
}

export function AlertBox(props: Props) {
  const theme = useContext(ThemeContext)
  // If we don't normalize upfront then grommet auto-detects darkness of background, and decides to use ['alert-box-warning'].dark with ['alert-box-warning-background'].dark
  const color = normalizeColor(mapStatus[props.status].color, theme)

  return (
    <Box
      border={{
        color: color,
        side: 'left',
        size: '3px',
      }}
      background={{ color: mapStatus[props.status].background }}
      pad={{ horizontal: 'small' }}
    >
      <Box direction="row" gap="small" align="center" justify={props.center ? 'center' : 'start'}>
        {props.icon && <Text color={color}>{props.icon}</Text>}
        <Text weight="bold" size="12px" style={{ marginTop: 10, marginBottom: 10 }}>
          {props.children}
        </Text>
      </Box>
    </Box>
  )
}
