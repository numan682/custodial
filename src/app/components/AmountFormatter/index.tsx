/**
 *
 * AmountFormatter
 *
 */
import { selectTicker } from 'app/state/network/selectors'
import * as React from 'react'
import { memo } from 'react'
import { useSelector } from 'react-redux'
import { Text } from 'grommet/es6/components/Text'
import { StringifiedBigInt } from 'types/StringifiedBigInt'
import { formatBaseUnitsAsRose, formatWeiAsWrose } from 'app/lib/helpers'
import { NoTranslate } from '../NoTranslate'

export interface AmountFormatterProps {
  amount: StringifiedBigInt | null
  amountUnit?: 'baseUnits' | 'wei'
  minimumFractionDigits?: number
  maximumFractionDigits?: number
  hideTicker?: boolean
  size?: string
  smallTicker?: boolean
}

/**
 * Formats base unit amounts to ROSEs
 */
export const AmountFormatter = memo(
  ({
    amount,
    amountUnit = 'baseUnits',
    minimumFractionDigits,
    maximumFractionDigits,
    hideTicker,
    size,
    smallTicker,
  }: AmountFormatterProps) => {
    const ticker = useSelector(selectTicker)
    const isUsingBaseUnits = amountUnit === 'baseUnits'
    if (amount == null) return <span>-</span>

    const formatter = isUsingBaseUnits ? formatBaseUnitsAsRose : formatWeiAsWrose
    const amountString = formatter(amount, {
      minimumFractionDigits: minimumFractionDigits ?? 1,
      maximumFractionDigits:
        typeof maximumFractionDigits !== 'undefined' ? maximumFractionDigits : isUsingBaseUnits ? 15 : 18,
    })

    const tickerProps = smallTicker
      ? {
          size: 'xsmall',
          weight: 600,
          color: 'lightText',
        }
      : {}

    return (
      <span>
        <span>{amountString}</span>
        {!hideTicker && (
          <Text size={size} {...tickerProps}>
            <NoTranslate>{` ${ticker}`}</NoTranslate>
          </Text>
        )}
      </span>
    )
  },
)
