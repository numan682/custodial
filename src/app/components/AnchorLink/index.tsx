import { Anchor, AnchorExtendedProps } from 'grommet/es6/components/Anchor'
import { Link, LinkProps } from 'react-router-dom'
import React from 'react'

type AnchorLinkProps = LinkProps & AnchorExtendedProps

/**
 * Behaves like {@link Link} and styled like {@link Anchor}. Needed because:
 * - Link isn't styled but routes correctly
 * - Anchor is styled but does a full reload
 */
export const AnchorLink = (props: AnchorLinkProps) => {
  return <Anchor as={Link} {...props} />
}
