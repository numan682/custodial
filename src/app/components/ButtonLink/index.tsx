import { Button, ButtonExtendedProps } from 'grommet/es6/components/Button'
import { Link, To } from 'react-router-dom'
import React from 'react'

type ButtonLinkProps = Omit<ButtonExtendedProps, 'label'> & { label: string; to: To }

/**
 * Behaves like {@link Link} and styled like {@link Button}. Needed because:
 * - Link isn't styled but routes correctly
 * - Button is styled but can't navigate
 * - Wrapping Link around Button adds two focusable elements
 */
export const ButtonLink = (props: ButtonLinkProps) => {
  const { label, to, ...buttonProps } = props
  return (
    <Link aria-label={label} to={to}>
      <Button {...buttonProps} label={label} tabIndex={-1} />
    </Link>
  )
}
