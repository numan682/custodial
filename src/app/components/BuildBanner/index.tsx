import { Trans, useTranslation } from 'react-i18next'
import { Alert } from 'grommet-icons/es6/icons/Alert'
import { Info } from 'grommet-icons/es6/icons/Info'
import { Box } from 'grommet/es6/components/Box'
import { AlertBox } from 'app/components/AlertBox'
import { buildBannerZIndex } from '../../../styles/theme/elementSizes'
import { AnchorLink } from '../AnchorLink'
import styled from 'styled-components'
import { deploys } from '../../../config'

const StickyBanner = styled(Box)`
  position: sticky;
  top: 0;
  z-index: ${buildBannerZIndex};
`

export const BuildBanner = () => {
  const { t } = useTranslation()

  if (window.location.origin === deploys.extension) {
    return null
  }

  if (window.location.origin === deploys.production) {
    return (
      <StickyBanner>
        <AlertBox status="ok" center icon={<Info size="20px" color="currentColor" />}>
          <Trans
            i18nKey="banner.domainMoved"
            t={t}
            components={{
              NewLink: (
                <AnchorLink to="https://wallet.oasis.io/" label="https://wallet.oasis.io/" color="text" />
              ),
            }}
            defaults="Oasis Wallet has a new home at <NewLink />."
          />
        </AlertBox>
      </StickyBanner>
    )
  }

  if (window.location.origin === deploys.staging) {
    return (
      <StickyBanner>
        <AlertBox status="warning" center icon={<Alert size="20px" color="currentColor" />}>
          {t(
            'banner.buildStaging',
            'Please note this is the staging deployment of Oasis Wallet. ONLY USE IT FOR TESTING.',
          )}
        </AlertBox>
      </StickyBanner>
    )
  }

  return (
    <StickyBanner>
      <AlertBox status="warning" center icon={<Alert size="20px" color="currentColor" />}>
        {t(
          'banner.buildPreview',
          'Please note this is an experimental build of Oasis Wallet and your secrets are not safe. ONLY USE IT FOR TESTING.',
        )}
      </AlertBox>
    </StickyBanner>
  )
}
