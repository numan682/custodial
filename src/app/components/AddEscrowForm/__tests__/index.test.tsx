import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { ModalProvider } from 'app/components/Modal'
import { Validator } from 'app/state/staking/types'
import * as React from 'react'
import { Provider } from 'react-redux'
import { configureAppStore } from 'store/configureStore'
import { transactionActions } from 'app/state/transaction'
import { WalletErrors } from 'types/errors'
import { AddEscrowForm } from '..'
import { ThemeProvider } from '../../../../styles/theme/ThemeProvider'

const renderComponent = (store: any, address: string, validatorStatus: Validator['status']) =>
  render(
    <Provider store={store}>
      <ThemeProvider>
        <ModalProvider>
          <AddEscrowForm validatorAddress={address} validatorStatus={validatorStatus} validatorRank={21} />
        </ModalProvider>
      </ThemeProvider>
    </Provider>,
  )

describe('<AddEscrowForm />', () => {
  let store: ReturnType<typeof configureAppStore>

  beforeEach(() => {
    store = configureAppStore()
  })

  it('should match snapshot', () => {
    const component = renderComponent(store, 'dummy-address', 'active')
    expect(component.container.firstChild).toMatchSnapshot()
  })

  it('should dispatch an addEscrow transaction', async () => {
    const spy = jest.spyOn(store, 'dispatch')
    renderComponent(store, 'dummy-address', 'active')

    await userEvent.type(screen.getByTestId('amount'), '1000')
    await userEvent.click(screen.getByRole('button'))

    expect(spy).toHaveBeenCalledWith({
      payload: {
        amount: '1000000000000',
        type: 'addEscrow',
        validator: 'dummy-address',
      },
      type: transactionActions.addEscrow.type,
    })
  })

  it('should render error indicator', () => {
    const storeWithError = configureAppStore({
      transaction: {
        active: true,
        error: {
          code: WalletErrors.NoOpenWallet,
          message: 'message',
        },
        success: false,
      },
    })

    renderComponent(storeWithError, 'dummy-address', 'active')

    expect(screen.getByText('errors.noOpenWallet')).toBeInTheDocument()
  })

  it('should warn before addEscrow to inactive validator', async () => {
    renderComponent(store, 'dummy-address', 'inactive')

    await userEvent.type(screen.getByTestId('amount'), '1000')
    await userEvent.click(screen.getByRole('button'))

    expect(screen.getByText('account.addEscrow.confirmDelegatingToInactive.title')).toBeInTheDocument()
  })
})
