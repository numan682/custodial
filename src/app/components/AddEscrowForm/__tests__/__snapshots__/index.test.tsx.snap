// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddEscrowForm /> should match snapshot 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  padding-top: 12px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  background-color: #FFFFFF;
  color: #16213E;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c5 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  width: 12px;
}

.c6 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #0092f6;
  border-radius: 4px;
  color: #16213E;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  background-color: #0092f6;
  color: #F8F8F8;
  border-radius: 4px;
  -webkit-transition-property: color,background-color,border-color,box-shadow;
  transition-property: color,background-color,border-color,box-shadow;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
  font-weight: bold;
}

.c6:hover {
  box-shadow: 0px 0px 0px 2px #0092f6;
}

.c6:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c6:focus > circle,
.c6:focus > ellipse,
.c6:focus > line,
.c6:focus > path,
.c6:focus > polygon,
.c6:focus > polyline,
.c6:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c6:focus::-moz-focus-inner {
  border: 0;
}

.c6:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c6:focus:not(:focus-visible) > circle,.c6:focus:not(:focus-visible) > ellipse,
.c6:focus:not(:focus-visible) > line,.c6:focus:not(:focus-visible) > path,
.c6:focus:not(:focus-visible) > polygon,.c6:focus:not(:focus-visible) > polyline,
.c6:focus:not(:focus-visible) > rect {
  outline: none;
  box-shadow: none;
}

.c6:focus:not(:focus-visible)::-moz-focus-inner {
  border: 0;
}

.c4 {
  box-sizing: border-box;
  font-size: inherit;
  font-family: inherit;
  border: none;
  -webkit-appearance: none;
  background: transparent;
  color: inherit;
  width: 100%;
  padding: 11px;
  font-weight: 600;
  margin: 0;
  border: 1px solid rgba(0,0,0,0.33);
  border-radius: 4px;
}

.c4:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c4:focus > circle,
.c4:focus > ellipse,
.c4:focus > line,
.c4:focus > path,
.c4:focus > polygon,
.c4:focus > polyline,
.c4:focus > rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c4:focus::-moz-focus-inner {
  border: 0;
}

.c4::-webkit-input-placeholder {
  color: #AAAAAA;
}

.c4::-moz-placeholder {
  color: #AAAAAA;
}

.c4:-ms-input-placeholder {
  color: #AAAAAA;
}

.c4::-webkit-search-decoration {
  -webkit-appearance: none;
}

.c4::-moz-focus-inner {
  border: none;
  outline: none;
}

.c4:-moz-placeholder,
.c4::-moz-placeholder {
  opacity: 1;
}

.c3 {
  position: relative;
  width: 100%;
}

.c0 {
  font-family: Rubik;
  font-size: 18px;
  line-height: 24px;
  background-color: #ffffff;
  color: #16213E;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

@media only screen and (max-width:768px) {
  .c1 {
    padding-top: 6px;
  }
}

@media only screen and (max-width:768px) {
  .c5 {
    width: 6px;
  }
}

<div
  class="c0"
  style="min-height: 100vh;"
>
  <form>
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
        >
          <input
            autocomplete="off"
            class="c4"
            data-testid="amount"
            id="amount-id"
            min="0"
            name="amount"
            placeholder="common.amount"
            required=""
            step="any"
            type="number"
            value=""
          />
        </div>
      </div>
      <div
        class="c5"
      />
      <button
        class="c6"
        type="submit"
      >
        account.addEscrow.delegate
      </button>
    </div>
  </form>
</div>
`;
